### ISSUE TITLE
Remote leechagent not working for implicit admins

### ISSUE THREAD

**Issue by Danking555** [ufrisk/LeechCore#27]
In my domain environment there are different groups of local administrators, and I'm a part of one of these - so I must be local admin on remote machines implicitly.
If I run net localgroup Administrators - there are only groups, without explicit users, except locally.
I installed the leechagent remotely and ran memprocfs without success - it couldn't communicate via rpc with the remote looking through the logs it was access denied.
Only after I added my user to the local administrators explicitly on the remote machine - rpc communication to remote worked. But I had to load the pmem driver manually (although the driver was in the right folder before installing)
Adding a specific user or set of users to local admins is not an option in my case and I'd like to make memproc work remotely automatically ;)
Thanks in advance!

**Comment by ufrisk** [comment]
Hi, Thanks for reporting these two independent issues.

1. pmem not loading. I have replicated the same issue on 1st connect. On 2nd connect the driver seems to be working fine. I'll investigate why this is happening. A short term workaround would be to connect twice.

2. admins in nested groups not working. If I put the admin directly in the Local Administrators group it's working. If I put the admin in an AD group which resides in Local Administrators group it's working. If I put the admin in a nested group I can confirm it's not working. I'll look into this, but things are a bit busy right now so it won't be this week probably.

**Comment by ufrisk** [comment]
I updated the LeechCore sources and binaries with bug fixes for these two issues.

Please let me know if it's working and I'll close the issue.

Also if you have any questions or feature requests around this please let me know.

**Comment by vintorxca** [comment]
First of all thanks for the quick responses and the update!
Secondly, I'm a local administrator on the remote, but unfortunately the driver is not loaded automatically.
The files that I put in the leechagent.exe directory before installing are:
The plugins directory
Python directory - with the embedded python files
info.db
leechagent.exe
leechcore.exe
vcruntime140.dll
vmm.dll
winpmem_x64.sys
And the I'm running leechagent.exe -remoteinstall <computer>

The error when running memprocfs: memprocfs.exe -remote rpc://<computer>$@<domain>:<computer>.<domain>
RPC: ERROR: Unable to open remote device #2 'pmem'
MemProcFS: Failed to connect to memory acquisition device

Am I missing something or what logs can I look out to help resolve this? Thanks

**Comment by Danking555** [comment]
First of all thanks for the quick responses and the update!
Secondly, I'm a local administrator on the remote, but unfortunately the driver is not loaded automatically.
The files that I put in the leechagent.exe directory before installing are:
The plugins directory
Python directory - with the embedded python files
info.db
leechagent.exe
leechcore.exe
vcruntime140.dll
vmm.dll
winpmem_x64.sys
And the I'm running leechagent.exe -remoteinstall

The error when running memprocfs: memprocfs.exe -remote rpc://computer$@domain:computer.domain
RPC: ERROR: Unable to open remote device #2 'pmem'
MemProcFS: Failed to connect to memory acquisition device

Am I missing something or what logs can I look out to help resolve this? Thanks

**Comment by Danking555** [comment]
First of all thanks for the quick responses and the update!
Secondly, I'm a local administrator on the remote, but unfortunately the driver is not loaded automatically - even when I'm trying to install/uninstall leechagent twice or more.

The files that I put in the leechagent.exe directory before installing are:
The plugins directory
Python directory - with the embedded python files
info.db
leechagent.exe
leechcore.exe
vcruntime140.dll
vmm.dll
winpmem_x64.sys
And the I'm running leechagent.exe -remoteinstall

The error when running memprocfs: memprocfs.exe -remote rpc://computer$@domain:computer.domain
RPC: ERROR: Unable to open remote device #2 'pmem'
MemProcFS: Failed to connect to memory acquisition device

Am I missing something or what logs can I look out to help resolve this? Thanks

**Comment by ufrisk** [comment]
Many thanks for looking into this 👍  Just wish to make sure the authentication issue has been resolved and that it's no longer an issue?

As for the winpmem issue it looks like it may still be around or is a completely other issue from the one I thought it was.

As for MemProcFS I'm guessing you're running it with the `-device pmem` as well. i.e.  `memprocfs.exe -device pmem -remote rpc://computer$@Domain:computer.domain` or similar.

---

1. can you try to run the MemProcFS twice. To see if it works on 2nd attempt if not on 1st. I had a bug like that in the past. I'm pretty sure it's not the issue though. But it should be an easy enough test.

2. on the remote host; are you able to run [most recent PCILeech](https://github.com/ufrisk/pcileech/releases/tag/v4.12) with winpmem_x64.sys in the same folder in an elevated administrative command prompt with the command `pcileech.exe display -min 0x1000 -device pmem -v -vv`

This should hopefully give some useful error messages if it's related to the winpmem driver loading. Also the event log may show some interesting artifacts for errors in driver loading. I'm not totally sure which event it is though.

**Comment by Danking555** [comment]
Yes, it seems that the issue with admins in nested groups is now working :)
The error I mentioned before is displayed when I tried to run the following command line:
`memprocfs.exe -device pmem -remote rpc://computer$@Domain:computer.domain`
It worked again after loading the driver manually on the remote machine and running this command again.
I'll look into the eventlogs and run pcileech when I'll be back at work and I'll update this comment

**Comment by Danking555** [comment]
However, the error I mentioned before is displayed when I tried to run the following command line:
`memprocfs.exe -device pmem -remote rpc://computer$@Domain:computer.domain`
I confirmed that the leech service is running on the remote machine and then ran the command - `pcileech display -min 0x1000 -device pmem -remote rpc://computer$@Domain:computer.domain -v -vv`
and the results were:
LeechRPC_RpcInitialize: 'iid_value@ncacn_ip_tcp:computer.domain[28473]'
RPC: ERROR: Unable to open remote device #2 'pmem'
PCILEECH: Failed to connect to memory acquisition device

What's very strange is that by looking at the eventlogs I can see event id 7045 about the service pmem that's been installed, but running `sc query pmem` indicates that the service doesn't exist.
Running `driverquery` doesn't list the pmem driver either.

Anyways, It worked again after loading the driver manually on the remote machine and running this command again.

**Comment by ufrisk** [comment]
I've looked more into this and I can't replicate the issue. It works fine for me unfortunately.

That the service isn't there when you look at it isn't very surprising. If the driver loading fails it will try to remove the service (if it was also created by the leechcore).

What are your system stats? windows version? physical/virtual machine? any virtualization based security features or anything special running?

**Comment by Danking555** [comment]
**I've finally figured it out!!!**
tldr; I changed the winpmem driver for the leechagent and it worked 💯 
There is no virtualization based security, the machine is physical, and it's one of the latest windows 10 versions.

**How I figured it out -** I loaded the winpmem driver with the latest winpmem_mini.exe release : `winpmem_mini.exe -l`
and noticed that there are many drivers that this command put into the temp directory, named: pme....tmp. 
Looking closely at their sizes I saw two different sizes - one of 45kb and another 46kb.
Using the driver sized 45kb doesn't work
But using the 46kb works fine! 👍  
Now I'm looking into using it
**Thanks for your desire to help and your work contribution to all these projects(memprocfs,pcileech,leechcore), I appreciate it!**

**Comment by AndrejSlavik** [comment]
Hey,
I am actually facing the similar problem as mentioned here with the WinPmem driver. 
Trying to recreate the situation from your video: https://youtu.be/Mij6LY1z4SY, 
but I always get stuck on this error: RPC: ERROR: Unable to open remote device #2 'pmem'.

I am using the latest WinPmem driver .sys file from official github in this link: https://github.com/Velocidex/WinPmem/releases/tag/v4.0.rc1

I have no idea what i am doing wrong here, I pasted the WinPmem driver directly to agent as shown in the video/manual, but still having trouble with this. I also tried all the commands you recommended here to @Danking555.

I was just not sure about his implementation which created the .tmp files, because solution requires .sys file, or am I missing something?

Thanks in advance.

![Snímka obrazovky 2023-04-10 200326](https://user-images.githubusercontent.com/79134197/230964296-54353a0b-7271-49ca-a3b7-280a89bdef80.png)
![Snímka obrazovky 2023-04-10 195841](https://user-images.githubusercontent.com/79134197/230964298-9d8815e8-ff24-4a7e-87a1-0e9769971647.png)
![Screenshot 2023-04-10 195958](https://user-images.githubusercontent.com/79134197/230964304-1a1cb75b-e36b-401c-88a0-9b162ef2e17e.png)

**Comment by ufrisk** [comment]
@AndrejSlavik In the syntax you use the pmem driver has to be local to `pcileech.exe` since you're not specifying the remote command. Put winpmem_x64.sys next to PCILeech, start from an elevated admin command prompt window `pcileech.exe display -min 0x1000 -device pmem -v -vv`.

Alternatively, if you wish to communicate with the LeechAgent service installed (and your computer is an AD connected computer) do:

`pcileech.exe display -min 0x1000 -device pmem -v -vv -remote rpc://<SPN>:localhost`  where <SPN> = your_computer_name$@your_ad_fqdn. If your computer is named BOB and the AD is AD.CONTOSO.COM then the SPN is BOB$@AD.CONTOSO.COM.

**Comment by ufrisk** [comment]
@AndrejSlavik what does the event log say, I think either Application Log or System Log may sometimes have entries related to winpmem such as failed to load driver which may shed some additional light on the issue.

Also make sure it's this driver you have downloaded https://github.com/Velocidex/WinPmem/raw/master/kernel/binaries/winpmem_x64.sys and that it starts with the characters MZ if you open it up in an hex editor or notepad. (It looks correct, but I've had others downloading some html off github and tried to load it as a driver - which obviously doesn't work, just trying to rule out some stupid things here).

**Comment by AndrejSlavik** [comment]
@ufrisk I tried to clean reinstall the whole agent with `leechagent.exe -remoteuninstall <host>`, made sure that every bit of LeechAgent was deleted from Program Files. After that command I additionally had to remove Python and Plugins directory because it still wasn't deleted. Also had to `sc stop pmem` to delete `winpmem_x64.sys` from the same folder to start over. 

Made sure to download the driver from link you provided, it started with the MZ. Copied to the LeechCore library again and replaced the old one and tried to `leechagent.exe -remoteinstall <host>` it again and the only logs I got from the install are these: 
![Screenshot 2023-04-11 234155](https://user-images.githubusercontent.com/79134197/231296212-e7e75a10-5126-4186-8ecb-4b2b9d1bce1b.png)
![Screenshot 2023-04-11 234227](https://user-images.githubusercontent.com/79134197/231296215-320f5cd0-ede7-4367-97fb-f465e6ae7196.png)
![Screenshot 2023-04-11 232424](https://user-images.githubusercontent.com/79134197/231296315-f7a69bba-51e7-4bbc-a880-fb9d9c12443d.png)

Also got this warning few moments later but don't know if it has any correlation whatsoever with the software import. 

![Screenshot 2023-04-11 234126](https://user-images.githubusercontent.com/79134197/231296656-381d0ac1-4023-4085-a021-83518f00cee5.png)

Could somehow AD settings be connected to this winpmem error?

**Comment by ufrisk** [comment]
The last issue does not seems to be related.

I believe this to be most likely a winpmem error and not directly related to the agent as such. Is it working if you start from an administrator command prompt similar to the one I have in the screenshot below? Also do you see something around the winpmem service in the System log?

![image](https://user-images.githubusercontent.com/18221464/231518443-1a86ee39-75b4-47e5-b612-369b1e441082.png)

**Comment by AndrejSlavik** [comment]
It actually worked on first try, then I tried to type in the command from the video but it just froze in the process of fetching the data from remote host. Then I repeated the command and again the error appeared once again. 

![Screenshot 2023-04-12 234259](https://user-images.githubusercontent.com/79134197/231897251-1053f308-de32-4fbe-a0d6-41a85c1e8831.png)

I couldn't simulate it again, I tried the command on your screenshot few more times but there was again this communication error with winpmem driver not quite sure how it's possible when it worked one time but moments later getting the same error.

**Comment by ufrisk** [comment]
I clearly have to look into this.

It may be that winpmem is having issue being loaded and reloaded too often as is the case with pcileech (since it's loaded and then unloaded for each command you run) on the target. You can also try to "mount" the memory with MemProcFS even doing the analysis on the remote computer and just forwarding the file system commands.

`memprocfs.exe -device pmem -remote rpc://...`  (memory analysis on your computer, pull the memory over the network).
`memprocfs.exe -device pmem -remotefs -remote rpc://...`  (memory analysis on the remote computer, file operations over the network).

I just released some major upgrades to the memprocfs project and now the plan is to look into the LeechCore project to see if I can add some performance and also stability to the winpmem and the rpc connection. Also I'll see if I can implement the LeechAgent as a powershell endpoint - which may ease some management for those that already have the WinRM port open...

**Comment by AndrejSlavik** [comment]
@ufrisk I tried to run the new update of MemProcFS remotely but still having issues with pmem driver
![Snímka obrazovky 2023-04-22 123158](https://user-images.githubusercontent.com/79134197/233792480-2f10ac1c-5ce2-4a61-8fb7-92bdd79db971.png)
![Screenshot 2023-04-22 171045](https://user-images.githubusercontent.com/79134197/233792560-f756ae45-0822-4de7-8972-c813a819bacf.png)

**Comment by ufrisk** [comment]
Hi, unfortunately I haven't had the time to find out what is causing these issues. It mostly work for me, but it too fail some times until I reboot the target computer.

The update was for other fixes and features so nothing have changed around this.

**Comment by ufrisk** [comment]
@AndrejSlavik I've recently made an update around the remote functionality. Not specifically to the winpmem driver but there are large changes otherwhere so it may be worthwile to try it again.

Especially I've made huge updates around the -remotefs option which should now be much more flexible and support all MemProcFS options. More info at: https://github.com/ufrisk/MemProcFS/wiki/_Remoting
